HDDS-6235. Empty KeyValueContainers are replicated without chunks directory.

errose28 opened the pull request:
## What changes were proposed in this pull request?

- Include empty chunks directory in the tarball when replicating a container.
- Recreate directory entries when unpacking the tarball.
- On datanode startup, create an empty chunks directory if no chunks directory is present.
    - Repairs existing containers now that the bug is fixed.

## What is the link to the Apache JIRA

HDDS-6235

## How was this patch tested?

- Unit test added, which passed only after the fix was applied.
- Manually tested chunks directory creation on startup in docker:
    1. Create empty container with `ozone admin container create`
    2. Delete its chunks directory.
    3. Restart the datanode.
    4. Check that empty chunks directory is created.

guihecheng commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:
```diff
@@ -254,6 +255,13 @@ public static void parseKVContainerData(KeyValueContainerData kvContainerData,
+      // If the container is missing a chunks directory, possibly due to the
+      // bug fixed by HDDS-6235, create it here.
+      File chunksDir = new File(kvContainerData.getChunksPath());
+      if (!chunksDir.exists()) {
```
Hi ethan, as I read, the chunks directory should be created in the TarContainerPacker while unpacking, then is there a case that we have to create this directory again in parseKVContainerData?

errose28 replied:
Hi @guihecheng, thanks for checking out the PR! There's two cases we want to fix here:
1. Empty containers that may need to be replicated in the future.
    - ReplicationManager is supposed to delete these, so this is more of a defensive code path.
2. Existing empty containers that got replicated before this patch and HDDS-5359 and lost their chunks directory.

Most of this patch is aimed at the first case, but these lines are aimed at the second case. We want to repair the existing containers that already lost their chunks directory so they do not get marked unhealthy by the scrubber (if running). Also, a container missing its chunks directory will fail replication on the datanode even before this patch, so SCM constantly reports it as under-replicated. There may be a lot of containers in this state due to the issue reported in HDDS-5359, which prevented many empty containers from being deleted.

errose28 replied:
parseKVContainerData is called on container import and startup. We are targeting the startup run to fix all the containers after the patch is applied. When this is called on import it should be a no-op since the chunks dir is now traveling with the tar as you mentioned.

guihecheng replied:
Oh, thanks for the explanation, actually we've seen similar container replicas with no "chunks" directory in our cluster, and I think this patch fixes the problem indeed.
LGTM+1

nandakumar131 reviewed (approved):
LGTM. Thanks @errose28 for working on this.

errose28 commented:
Thanks for reviews @nandakumar131 @hanishakoneru. After some offline discussion we decided to remove the metadata scan on import/export and the unhealthy mark if there's an issue with the container on export. I filed HDDS-6344 for gentler handling of incorrect but recoverable containers, which could be used to flag container content errors on import/export.

errose28 commented:
After this patch, missing chunks directories will be recreated in `KeyValueContainerData#parseKVContainerData`. This means the test in the recently merged metadata inspector patch to fix missing chunks directories fails and is not needed. I removed the metadata inspector test, and replaced it with a new unit test that makes sure missing chunks directories are created when `KeyValueContainerData#parseKVContainerData` is called.

hanishakoneru commented:
LGTM. +1 pending CI.
Unit test failure is unrelated. Retriggering CI.

adoroszlai commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/TestKeyValueContainer.java:
```diff
@@ -244,6 +268,18 @@ public void testContainerImportExport() throws Exception {
     }
   }
 
+  private void checkContainerFilesPresent(KeyValueContainerData data,
```
Nit: this method seems to be unused.

adoroszlai commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/TarContainerPacker.java:
```diff
@@ -84,11 +84,19 @@
           Path destinationPath = dbRoot
               .resolve(name.substring(DB_DIR_NAME.length() + 1));
-          extractEntry(archiveInput, size, dbRoot, destinationPath);
+          if (entry.isDirectory()) {
+            Files.createDirectories(destinationPath);
+          } else {
+            extractFileEntry(archiveInput, size, dbRoot, destinationPath);
+          }
```
Nit: can we move this logic to `extractEntry` (and keep the method name) to avoid duplication?

errose28 replied:
done!

adoroszlai reviewed (commented):
Thanks @errose28 for the patch, LGTM.  I only have two minor suggestions.

Also, would you mind adding a unit test in `TestTarContainerPacker`?

errose28 commented:
Thanks for the review @adoroszlai. I was initially confused by your test comments, but then realized the merge had deleted my unit test, so that was a good catch. I restored the test so hopefully that part makes more sense now.

adoroszlai reviewed (approved):
Thanks @errose28 for updating the patch.

adoroszlai commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/TarContainerPacker.java:
```diff
@@ -109,27 +111,31 @@
+  private void extractEntry(ArchiveEntry entry, InputStream input, long size,
+      Path ancestor, Path path) throws IOException {
     HddsUtils.validatePath(path, ancestor);
     Path parent = path.getParent();
     if (parent != null) {
       Files.createDirectories(parent);
     }
```
Nit: we could move this inside the `else` branch, since in the `if` branch `createDirectories(path)` would take care of parent directories.

errose28 replied:
Yeah I agree that reads better. Done.

symious commented:
@errose28 Thanks for the patch, our cluster does have the same issue while doing replication containers.

One thing I noticed while running the unit test is even without the changes in TarContainerPacker, the test also passed. Does that mean the key point to this issue is the creation of the CHUNK_PATH in `KeyValueContainerUtil`? Co

symious commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/TarContainerPacker.java:
```diff
@@ -209,6 +216,12 @@ public void pack(Container<KeyValueContainerData> container,
+    archiveOutput.putArchiveEntry(entry);
+
+    // Add files in the directory.
     try (Stream<Path> dirEntries = Files.list(dir)) {
```
I think the error is thrown here, so it still needs the CHUNK file to exist?

Reproduced by adding `new File(data.getChunksPath()).delete();` after checking the state of original container.

hanishakoneru replied:
@symious the error is thrown because the chunk directory does not exist. Not having chunk files in the directory should not be a problem here.

hanishakoneru commented:
> One thing I noticed while running the unit test is even without the changes in TarContainerPacker, the test also passed. Does that mean the key point to this issue is the creation of the CHUNK_PATH in KeyValueContainerUtil?

KeyValueContainerUtil change fixes existing containers with missing chunks directory. This could have occurred when a container with empty chunks directory was replicated before. The change in TarContainerPacker is needed so that when containers with empty chunks directories are replicated, the replica container get the chunks directory even though it does not have any contents.

errose28 commented:
Hi @symious following up on your questions:
> Does that mean the key point to this issue is the creation of the CHUNK_PATH in KeyValueContainerUtil?

There's two parts to this PR: preventing the issue and fixing existing containers. I think [this comment](https://github.com/apache/ozone/pull/3052#discussion_r801921918) might answer the question.

> One thing I noticed while running the unit test is even without the changes in TarContainerPacker, the test also passed

I cannot repro this. Running master from 68c5ac5df4fbc0edd7114394112fa696a3dc9229 (before this patch was merged), I copy/pasted in the changes to TestKeyValueContainer and both new tests failed. Can you please check again?

symious commented:
@hanishakoneru Thanks for the explaination.

@errose28 I created a branch with the changes in TarContainerPacker removed here https://github.com/symious/ozone/tree/HDDS-6235, the unit test passed, could you help to check?

symious commented:
I think the process sequence are as follows:
1. container exported to FileOutputStream
2. peer use importContainerData to import the contaienr. 
3. TarContainerPacker unpack the container from the FileInputStream. After the unpack, the directory of "chunks" shouldn't be there.
4. But in `KeyValueContainerUtil.parseKVContainerData(containerData, config);` since the new logic was added that if the chunk directory doesn't exists, we create it.

So even without the changes in TarContainerPacker, the unit test still passed.
